python/ray/tune/api_v2/tuner.py

```diff
@@ -0,0 +1,345 @@
+import copy
+from dataclasses import dataclass
+import os
+import traceback
+from typing import Any, Callable, Dict, List, Optional, Type, Union
+
+import ray
+import ray.cloudpickle as pickle
+from ray.data import Dataset
+
+from ray.tune import TuneError, Experiment
+from ray.tune.api_v2.convertible_to_trainable import ConvertibleToTrainable
+from ray.tune.callback import Callback
+from ray.tune.schedulers import TrialScheduler
+from ray.tune.suggest import Searcher
+from ray.tune.trainable import Trainable
+from ray.tune.tune import run
+from ray.util import PublicAPI
+from ray.util.client.common import ClientActorHandle
+from ray.util.ml_utils.node import force_on_current_node
+
+
+########################################################################
+# The motivations behind the new Tuner API as part of unified ML effort is that
+# 1. Users can seamlessly transition from their training process to tuning process.
+# 2. Both training and tuning now use Tuner API as the execution engine.
+#   This will help consolidate code on both sides.
+# 3. Dataset and data preprocessing are now an integral part of training/tuning process.
+#   Bonus point: you can even tune dataset and preprocessing.
+
+# Some requirements on Tuner that may help you understand some implementation decisions:
+# 1. Tuner needs to work in ray client mode.
+# 2. Tuner needs to be restored in case of any failures.
+########################################################################
+
+
+@dataclass
+@PublicAPI
+class TuneAlgoConfig:
+    """Tune specific config, related to how hyperparameter tuning algorithms
+    are set up."""
+
+    # Currently this is not at feature parity with `tune.run`, nor should it be.
+    # The goal is to reach a fine balance between API flexibility and conciseness.
+    # We should carefully introduce arguments here instead of just dumping everything.
+    mode: Optional[str] = None
+    metric: Optional[str] = None
+    search_alg: Optional[Searcher] = None
+    scheduler: Optional[TrialScheduler] = None
+    num_samples: int = 1
+
+
+# The magic key that is used when instantiating Tuner during resume.
+_TUNER_INTERNAL = "tuner_internal"
+# The key that denotes the dataset config in param space.
+_DATASETS = "datasets"
+
+
+class _TunerInternal:
+    """The real implementation behind external facing ``Tuner``.
+
+    The external facing ``Tuner`` multiplexes between local Tuner and remote Tuner
+    depending on whether in Ray client mode.
+
+    In Ray client mode, external Tuner wraps ``_TunerInternal`` into a remote actor,
+    which is guaranteed to be placed on head node.
+    """
+
+    def __init__(
+        self,
+        restore_path: str = None,
+        trainable: Optional[
+            Union[
+                str,
```

Comment:
Can we drop str?

Reply:
hmmm, I personally don't like the idea of overloading too many possibilities in a single argument. However, people do like to use this for RLlib trainables, by, e.g., specifying `PPO`. 
Do you foresee a way that use can still enjoy this conveniency but not at the cost of a clear API?